solrmonitor/zkdispatcher.go

```diff
@@ -120,19 +117,34 @@ func (d *ZkDispatcher) eventLoop() {
 	defer atomic.AddInt32(&d.runningProcs, -1)
 
 	for {
-		chosen, recv, recvOK := reflect.Select(d.selectCases)
+		// first check if the dispatcher is closed or if there is a new handler
+		closing := false
+		select {
+		case nh := <- d.newHandlerChan:
+			d.selectHandlers = append(d.selectHandlers, nh.handler)
+			d.selectCases = append(d.selectCases, newCase(nh.watcher))
+		case <- d.closedChan:
+			closing = true
+			d.logger.Printf("zkdispatcher: exiting via close channel")
+		default:
+		}
```

Comment:
I think I'd put this select into its own loop, as long as you can keep successfully drained the incoming channel I'd keep doing it, since the actual workers are all running in the background and could be shoving new watches in "rapidly".  I'm imagining a BFS tree watch where discovering a single node node causes you watch all of its children.